utils/bimap/bimap.go

```diff
@@ -0,0 +1,169 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package bimap
+
+import (
+	"sync"
+
+	"github.com/ava-labs/avalanchego/utils"
+)
+
+var _ BiMap[int, int] = (*biMap[int, int])(nil)
+
+type Entry[K, V any] struct {
+	Key   K
+	Value V
+}
+
+// BiMap is a bi-directional map.
+type BiMap[K, V any] interface {
+	// Put the key value pair into the map. If either [key] or [val] was
+	// previously in the map, the previous entries will be removed and returned.
+	//
+	// Note: Unlike normal maps, it's possible that Put removes 0, 1, or 2
+	// existing entries to ensure that mappings are one-to-one.
+	Put(key K, val V) (removed []Entry[K, V])
+	// GetKey that maps to the provided value.
+	GetKey(val V) (key K, exists bool)
+	// GetValue that is mapped to the provided key.
+	GetValue(key K) (val V, exists bool)
+	// HasKey returns true if [key] is in the map.
+	HasKey(key K) (exists bool)
+	// HasValue returns true if [val] is in the map.
+	HasValue(val V) (exists bool)
+	// DeleteKey removes [key] from the map and returns the value it mapped to.
+	DeleteKey(key K) (val V, deleted bool)
+	// DeleteValue removes [val] from the map and returns the key that mapped to
+	// it.
+	DeleteValue(val V) (key K, deleted bool)
+	// Inverse the keys and values of this map.
+	//
+	// Note: The returned map will modify the same underlying state.
+	Inverse() BiMap[V, K]
+	// Len return the number of entries in this map.
+	Len() int
+}
+
+type biMap[K, V comparable] struct {
+	lock       *sync.RWMutex
```

Comment:
Should we just let the caller handle concurrency and get rid of this lock?

Reply:
ok